# 01. Importing libraries

In [1]:
# Import libraries
import warnings
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

from pandas.core.common import SettingWithCopyWarning
from scipy.stats import norm

# 02. Importing data

In [2]:
# Define csv path
path = r'C:\Users\lucav\Desktop\Instacart Basket Analysis'

In [3]:
# Import data from 'orders_products_merged_frequencies&flags.pkl' and 'customers.csv'
warnings.simplefilter(action='ignore', category=FutureWarning) # Added to ignore warnings
ords_prods = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_frequencies&flags.pkl'))
df_customers = pd.read_csv(os.path.join(path, '02 Data','Original Data', 'customers.csv'))

# 03. Data wrangling

## Wrangle data in 'df_customers' so that it follows consistent logic

In [4]:
# Print the first 5 rows from 'df_customers'
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
# Rename 'First Name' column to 'first_name', 'Surnam' to 'last_name', 'Gender' to 'gender', 'STATE' to 'state', 'Age' to 'age'
df_customers.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age'}, inplace = True)

In [6]:
# Print the first 5 rows from 'df_customers' to ensure changes were properly applied
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# 04. Data quality and consistency checks

## Data type

In [7]:
# Check the data type of each column in 'df_customers'
df_customers.dtypes

user_id          int64
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

The 'user_id' column can be changed from int64 type to string type.

In [8]:
# Change 'user_id' type to string
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [9]:
# Verify that 'user_id' type has been changed to string
df_customers.dtypes

user_id         object
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [10]:
# Check for mixed-type data in 'df_customers'
for col in df_customers.columns.tolist():
    weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_customers[weird]) > 0:
        print (col)

first_name


The 'first_name' column contains mixed-type data but it should be a string type. This requires additional checks to identify which values are stored differently.

In [11]:
# Return only those rows where the 'first_name' column is not a string
df_customers[df_customers['first_name'].apply(lambda x: not isinstance(x, str))]

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

It can be observed that 11259 rows have 'NaN' values in the 'first_name' column, that are not stored as a string type. This issue would need to be discussed with stakeholders to better assess the impact on the analysis.

## Missing values

In [12]:
# Find missing values in 'df_customers'
df_customers.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [13]:
# Check the missing values in the 'df_customers' column
df_customers[df_customers['first_name'].isnull()==True]

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

As previously discovered, 11259 values are missing in the 'first_name' column and the issue should be discussed with stakeholders. At the moment, data won't be modified because the missing values aren't deemed vital to the analysis.

## Duplicate values

In [14]:
# Find duplicate values in 'df_customers'
df_customers_dups = df_customers[df_customers.duplicated()]

# Return 'df_customers_dups'
df_customers_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

No duplicate values found.

## Look for anomalies

In [15]:
# Run the describe() function on 'df_customers'
df_customers.describe()

age   n_dependants         income
count  206209.000000  206209.000000  206209.000000
mean       49.501646       1.499823   94632.852548
std        18.480962       1.118433   42473.786988
min        18.000000       0.000000   25903.000000
25%        33.000000       0.000000   59874.000000
50%        49.000000       1.000000   93547.000000
75%        66.000000       3.000000  124244.000000
max        81.000000       3.000000  593901.000000

No outstanding anomalies observed.

# Export 'df_customers' as 'customers_prepared.pkl'

In [16]:
df_customers.to_pickle(os.path.join(path,'02 Data','Prepared Data','customers_prepared.pkl'))

# 05. Combine 'df_customers' with 'ords_prods'

In [17]:
# Print the first 5 rows from 'df_customers'
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [18]:
# Print the first 50 rows from 'df_customers'
ords_prods.head(50)

order_id  user_id  order_number  day_of_week  hour_of_day  \
0    2539329        1             1            2            8   
1    2398795        1             2            3            7   
2     473747        1             3            3           12   
3    2254736        1             4            4            7   
4     431534        1             5            4           15   
5    3367565        1             6            2            7   
6     550135        1             7            1            9   
7    3108588        1             8            1           14   
8    2295261        1             9            1           16   
9    2550362        1            10            4            8   
10   2968173       15            15            1            9   
11   1870022       15            17            2           16   
12   1911383       15            18            2           11   
13   2715276       15            21            1            9   
14    487368       15            22            1           10   
15   2293453       19             2            5           14   
16   1973799       19             5            6           12   
17    532817       19             7            4           17   
18   1573906       21            10            3           10   
19   1593000       31            10            3            8   
20   2231262       31            17            3           11   
21   2580647       43             6            4           16   
22   2187180       43             9            4           12   
23   2497897       52             1            1            9   
24   1318871       52             2            1           10   
25   1261384       52             3            1           10   
26    580568       52             5            2           10   
27   2428073       52             7            2            7   
28    342306       52             9            1           11   
29   1520353       52            10            2           10   
30   2542086       52            11            2            9   
31    944694       52            12            1           16   
32   1268191       52            15            2           11   
33   1122089       52            17            2           10   
34   1498922       52            18            2           10   
35    180919       52            19            3            8   
36   2695875       52            21            5           10   
37   3205960       67             1            3           11   
38    970922       67             2            1           15   
39    596585       67             3            1           13   
40   2037087       67             4            2            9   
41   2474111       67             5            3           14   
42   1568481       67             6            3           12   
43   2048527       67             8            2           11   
44   2389031       67             9            2           14   
45   2378562       67            10            1           10   
46   1515698       67            11            1           11   
47   1928050       67            12            2           12   
48   1949274       67            13            1           10   
49    551869       67            14            1            9   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                     15.0        False         196                  1   
2                     21.0        False         196                  1   
3                     29.0        False         196                  1   
4                     28.0        False         196                  1   
5                     19.0        False         196                  1   
6                     20.0        False         196                  1   
7                     14.0        False         196                  2   
8                     

The key column that will be used to merge 'df_customers' with 'ords_prods' is '**user_id**'.

In [19]:
# Check the data types of 'ords_prods'
ords_prods.dtypes

order_id                    int64
user_id                     int64
order_number                int64
day_of_week                 int64
hour_of_day                 int64
days_since_prior_order    float64
first_order                  bool
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
price_range_loc            object
busiest_days               object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
customer_spending         float64
spending_flag              object
order_frequency           float64
order_frequency_flag       object
dtype: object

In [20]:
# Change the type of 'user_id' column in 'ords_prods' to string
ords_prods['user_id'] = ords_prods['user_id'].astype('str')

In [21]:
# Check that the data type of 'user_id' is the same in 'df_customer' and 'ords_prods'
df_customers['user_id'].dtype == ords_prods['user_id'].dtype

True

In [22]:
# Merge 'df_customers' with 'ords_prods' using 'user_id' as key column
df_ords_prods_cust = ords_prods.merge(df_customers, on = 'user_id')

In [23]:
# Print the first 5 rows from 'df_ords_prods_cust'
df_ords_prods_cust.head()

order_id user_id  order_number  day_of_week  hour_of_day  \
0   2539329       1             1            2            8   
1   2398795       1             2            3            7   
2    473747       1             3            3           12   
3   2254736       1             4            4            7   
4    431534       1             5            4           15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ...   order_frequency_flag  first_name  last_name  gender  \
0          0  ...  Non-frequent customer       Linda     Nguyen  Female   
1          1  ...  Non-frequent customer       Linda     Nguyen  Female   
2          1  ...  Non-frequent customer       Linda     Nguyen  Female   
3          1  ...  Non-frequent customer       Linda     Nguyen  Female   
4          1  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age date_joined  n_dependants fam_status  income  
0  Alabama  31   2/17/2019             3    married   40423  
1  Alabama  31   2/17/2019             3    married   40423  
2  Alabama  31   2/17/2019             3    married   40423  
3  Alabama  31   2/17/2019             3    married   40423  
4  Alabama  31   2/17/2019             3    married   40423  

[5 rows x 32 columns]

# 06. Export 'df_ords_prods_cust' as 'orders_products_customers_merged.pkl'

In [24]:
df_ords_prods_cust.to_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_customers_merged.pkl'))